openedx/core/djangoapps/xblock/README.rst

```diff
@@ -0,0 +1,94 @@
+  - Runtime code and services which can be shared among all instances of the runtime are encapsulated in a new ``XBlockRuntimeSystem`` singleton, again for memory and performance reasons.
+* `learning_context <./learning_context/>`_ : **Learning Context** management code: defines the abstract base class for learning contexts, the base opaque keys that learning contexts use, and provides a plugin-based API to get the learning context given any usage key. See `Learning Contexts`_ below.
+
+Learning Contexts
+-----------------
+
+A "Learning Context" is a course, a library, a program, or some other collection of content where learning happens.
+
+To work with this runtime, all learnable XBlock content for a learning context must ultimately be stored in a Blockstore bundle, but the runtime does not impose any restrictions beyond that. The relationship between a learning context and a bundle will usually be 1:1, but some future types of learning context may not stick to that, and the runtime doesn't require it.
+
+Currently only `the "Content Library" learning context <./learning_context/content_library/>`_ is implemented. Additional learning contexts can be implemented as plugins, by subclassing the ``LearningContextImpl`` class and registering in the ``openedx.learning_context`` entry point.
+
+A learning context is responsible for:
+
+* Definining whether or not a given XBlock [usage key] exists in the learning context (e.g. "Does the content library X contain block Y?")
+
+  - This is done via its ``definition_for_usage()`` method which must return ``None`` if the usage key doesn't exist in the context.
+  - There is **no** general method to "list all XBlocks" in the learning context, because it's expected that learning contexts may be dynamic, e.g. with content assigned just in time via adaptive learning. However, if a learning context is static it can certainly implement an API to list all the blocks it contains.
+* Determining whether or not a given user has **permission to view/use** a given XBlock and/or to **edit** that XBlock.
+
+  - This is done by via its ``can_view_block()`` and ``can_edit_block()`` methods.
+  - For example, "pathways" might allow any user to view any XBlock, but "courses" require enrollment, cohort, and due date checks as part of this permissions logic.
+
+* Mapping usage keys to ``BundleDefinitionLocator`` keys.
+
+  - The XBlock runtime and other parts of the system do not know nor prescribe how a usage locator like ``lb:library15:html:introduction`` (HTML block with usage ID "introduction" in library with slug "library15") maps to an OLX file in Blockstore like "``html/introduction/definition.xml` in bundle with UUID `11111111-1111-1111-1111-111111111111`" - that logic is left to the learning context.
```

Comment:
BTW, this means that any files with ":" in them are illegal in Blockstore, right? (I have no objection to that, I just want to verify.)

Reply:
> BTW, this means that any files with ":" in them are illegal in Blockstore, right?

Not necessarily; for one thing, it's only in the opaque key format, so image files, video files, or other non-OLX things could include `:` in their paths. For another, the key parsing code I wrote only looks at the first N `:` instances, so you couldn't put a `:` in the first two parts (the library slug and the XBlock type), but you could in the last part, like:

    lb:library15:html:this:part:can:contain:lots:of:colons
        ^^^^^^^^ ^^^^
        no colons

That said, I think that it's easier just to say "no colons".

I'm totally open to using any separator, though I personally prefer `:` to `/` or `+` because it makes writing URL patterns easier and avoids URL encoding/decoding ambiguity. If anyone has strong opinions let's hear them sooner rather than later :)

Reply:
I have a mild preference for + because it's what we do elsewhere and because ":" is an illegal character in a lot of filesystems, and it seems like it would be useful to be able to make a folder with the key as the name. That being said, it's relatively simple to escape, so my commitment to it is only at bikeshed levels, and it sounds like you've run into more issues with + than I have.

Reply:
> because ":" is an illegal character in a lot of filesystems, and it seems like it would be useful to be able to make a folder with the key as the name.

Ah, but you'd have to remove the prefix because all opaque keys have a prefix like `block-v1:` which contains a colon, so you can never use an opaque key on a filesystem that doesn't allow colons anyways (and I haven't really seen people use serialized keys without the prefix, which I don't think we'd want to encourage).